swiftide-integrations/src/lancedb/retrieve.rs

```diff
@@ -124,6 +127,69 @@ impl Retrieve<SimilaritySingleEmbedding> for LanceDB {
     }
 }
 
+#[async_trait]
+impl Retrieve<CustomStrategy<query::VectorQuery>> for LanceDB {
+    /// Implements vector similarity search for LanceDB using a custom query strategy.
+    ///
+    /// # Type Parameters
+    /// * `VectorQuery` - LanceDB's query type for vector similarity search
+    async fn retrieve(
+        &self,
+        search_strategy: &CustomStrategy<query::VectorQuery>,
+        query: Query<states::Pending>,
+    ) -> Result<Query<states::Retrieved>> {
+        // Build the custom query using both strategy and query state
+        let query_builder = search_strategy.build_query(&query)?;
+
+        // Execute the query using the builder's built-in methods
+        let batches = query_builder
+            .execute()
+            .await?
+            .try_collect::<Vec<_>>()
+            .await?;
+
+        let mut documents = vec![];
+
+        for batch in batches {
+            let schema: SchemaRef = batch.schema();
+
+            for row_idx in 0..batch.num_rows() {
+                let mut metadata = Metadata::default();
+                let mut content = String::new();
+
+                for (col_idx, field) in schema.fields().iter().enumerate() {
+                    let column = batch.column(col_idx);
+
+                    if let Some(array) = column.as_any().downcast_ref::<StringArray>() {
+                        if field.name() == "chunk" {
+                            // Extract the "content" field
+                            content = array.value(row_idx).to_string();
+                        } else {
+                            // Assume other fields are part of the metadata
+                            let value = array.value(row_idx).to_string();
+                            metadata.insert(field.name().clone(), value);
+                        }
+                    } else {
+                        // Handle other array types as necessary
+                        // TODO: Can't we just downcast to serde::Value or fail?
+                    }
+                }
```

Comment:
Perhaps nice to refactor this to a `from_record_batch(batch) -> Documents`, so the code isn't duplicated across retriever implementations